arraycontext/impl/pytato/compile.py

```diff
@@ -0,0 +1,283 @@
+    def __eq__(self, other):
+        raise NotImplementedError
+
+    def __hash__(self):
+        raise NotImplementedError
+
+
+@dataclass(frozen=True, eq=True)
+class ScalarInputDescriptor(AbstractInputDescriptor):
+    dtype: np.dtype
+
+
+@dataclass(frozen=True, eq=True)
+class ArrayContainerInputDescriptor(AbstractInputDescriptor):
+    id_to_ary_descr: "PMap[Tuple[Any, ...], Tuple[np.dtype, \
+                                                  Tuple[int, ...]]]"
+
+
+def _ary_container_key_stringifier(keys: Tuple[Any, ...]) -> str:
+    """
+    Helper for :meth:`PytatoCompiledOperator.__call__`. Stringifies an
+    array-container's component's key. The aim is that no-two different keys
+    have the same stringification.
+    """
+    def _rec_str(key: Any) -> str:
+        if isinstance(key, (str, int)):
+            return str(key)
+        elif isinstance(key, tuple):
+            return "tup" + "_".join(_rec_str(k) for k in key) + "endtup"
+        else:
+            raise NotImplementedError
+
+    return "_".join(_rec_str(key) for key in keys)
+
+
+@dataclass
+class PytatoCompiledOperator:
+    """
+    Records a side-effect-free callable :attr:`PytatoCompiledOperator.f`, that
+    would be specialized for different input types
+    :meth:`PytatoCompiledOperator.__call__` is invoked with.
+
+    .. attribute:: f
+
+        The callable that would be specialized into :mod:`pytato` DAGs.
+
+    .. automethod:: __call__
+    """
+
+    actx: PytatoPyOpenCLArrayContext
+    f: Callable[..., Any]
+    program_cache: Dict[Tuple[AbstractInputDescriptor, ...],
+                        "PytatoExecutable"] = field(default_factory=lambda: {})
+
+    def __call__(self, *args: Any) -> Any:
+        """
+        Mimics :attr:`~PytatoCompiledOperator.f` being called with *args*.
+        Before calling :attr:`~PytatoCompiledOperator.f`, it is compiled to a
+        :mod:`pytato` DAG that would apply :attr:`~PytatoCompiledOperator.f`
+        with *args* in a lazy-sense.
+        """
+
+        from arraycontext.container.traversal import (rec_keyed_map_array_container,
+                                                      is_array_container)
+
+        def to_arg_descr(arg: Any) -> AbstractInputDescriptor:
+            if np.isscalar(arg):
+                return ScalarInputDescriptor(np.dtype(arg))
+            elif is_array_container(arg):
+                id_to_ary_descr = {}
+
+                def id_collector(keys, ary):
+                    id_to_ary_descr[keys] = (np.dtype(ary.dtype),
+                                             ary.shape)
+                    return ary
+
+                rec_keyed_map_array_container(id_collector, arg)
+                return ArrayContainerInputDescriptor(pmap(id_to_ary_descr))
+            else:
+                raise ValueError("Argument to a compiled operator should be"
+                                 " either a scalar or an array container. Got"
+                                 f" '{arg}'.")
```

Comment:
Could this function be moved outside? (Maybe with an underscore?)

Reply:
Yep, thanks: f618fa4.